### PR TITLE
Add support for allocating PodCIDRs from multiple IPAM pools

### PR DESCRIPTION
**Pull request by gandro**
This adds support for IPAM pools. More specifically, this adds a new IPAM mode (`multi-pool`) which allows nodes to allocate PodCIDRs from different pools, depending on properties of the workload as defined by the user. An example where this will be useful is multi-homing (see https://github.com/cilium/cilium/issues/20129). 

Notably, this PR only adds the IPAM infrastructure which allows consumers inside Cilium-Agent to allocate PodCIDRs from non-default pools, it does not yet add any consumers which actually make use of this feature. This means that this PR is not going to be very useful on its own. At the moment, we require user to statically configure which IPAM pools a node should allocate from via the `multi-pool-node-pre-alloc` flag. This way of configuring a node's IPAM pools will likely be superseded by a custom network configuration CRD.

Overview of the current mechanism:

Pooled IPAM CiliumNode CRD

As with other IPAM modes, cilium-operator is responsible for global allocation decisions, i.e. allocating PodCIDRs to nodes. Each cilium-agent indicates its own demand in the CiliumNode CRD. Once it has received a PodCIDR for a given pool, it will then hand out individual IPs based on pool specified in the IP allocation request (typically received from the CNI plugin).

*Known Limitations*:
- Encryption and tunnel mode are not supported (blocked on https://github.com/cilium/cilium/issues/20170)
- In the current form, this will not set up local node routes, since the setup of routes for alternative networks will be handled by the multi-homing feature instead. For now, we requires endpoint routes being enabled.
- Cilium's built-in legacy masquerading does not support multiple pod CIDRs (c.f. https://github.com/cilium/cilium/issues/22273). To work around this, use either BPF masquerading if there is only one single native-routing-cidr. Otherwise, use ip-masq-agent (iptables or BPF work both)
- This version of the PR requires that pools are statically defined in the operator config, using the newly added `ipam.operator.multiPoolMap` Helm value. Support for managing via CRDs instead will be implemented in a follow-up PR (c.f. [`ippool-crd`](https://github.com/tklauser/cilium/tree/pr/tklauser/ippool-crd) branch here)

Joint work with @tklauser

*Review per commit*

Meta-issue: #24764

### CODE REVIEW

**Review by zacharysarah — APPROVED** [review]
LGTM from a docs perspective.

**Comment by gandro** [comment]
Pushed a small commit to add support for `auto-direct-node-routes`: [323d01f](https://github.com/cilium/cilium/pull/22762/commits/323d01f70bed3630e61bd435f08593b5d468cf31). Other than that, no changes, just a rebase on main.

File: pkg/ipam/pool.go [hunk 1]
```diff
@@ -49,25 +47,15 @@ type podCIDRPool struct {
...
-func newPodCIDRPool(allocationThreshold, releaseThreshold int, previouslyReleasedCIDRs []string) *podCIDRPool {
-	if allocationThreshold <= 0 {
-		allocationThreshold = defaults.IPAMPodCIDRAllocationThreshold
-	}
-
-	if releaseThreshold <= 0 {
-		releaseThreshold = defaults.IPAMPodCIDRReleaseThreshold
-	}
-
+func newPodCIDRPool(previouslyReleasedCIDRs []string) *podCIDRPool {
```
Comment (squeed) [inline pkg/ipam/pool.go:46]: nit: the docblock is wrong now

File: pkg/ipam/pool.go [hunk 2]
```diff
@@ -31,12 +31,10 @@ import (
...
 type podCIDRPool struct {
-	mutex               lock.Mutex
-	ipAllocators        []*ipallocator.Range
-	released            map[string]struct{}
-	removed             map[string]struct{}
-	allocationThreshold int
-	releaseThreshold    int
+	mutex        lock.Mutex
+	ipAllocators []*ipallocator.Range
+	released     map[string]struct{}
```
Comment (squeed) [inline pkg/ipam/pool.go]: from discussion: it would be nice to know what the map keys are here.

File: pkg/ipam/clusterpool_v2beta2.go [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+		c.mutex.Unlock()
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.poolsUpdated:
+			continue
+		case <-timer.After(5 * time.Second):
+			log.WithFields(logrus.Fields{
+				logfields.HelpMessage: "Check if cilium-operator pod is running and does not have any warnings or error messages.",
```
Comment (squeed) [inline pkg/ipam/clusterpool_v2beta2.go]: add pool name here too

File: pkg/ipam/clusterpool_v2beta2.go [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+}
+
+func (c *clusterPoolManager) waitForPool(ctx context.Context, family Family, poolName string) {
+	timer, stop := inctimer.New()
+	defer stop()
+	for {
+		c.mutex.Lock()
+		switch family {
+		case IPv4:
+			if p, ok := c.pools[poolName]; ok && p.v4 != nil && p.v4.hasAvailableIPs() {
```
Comment (squeed) [inline pkg/ipam/clusterpool_v2beta2.go]: from meeting: what happens if the pool happened to be full on startup?

Comment (squeed) [reply pkg/ipam/clusterpool_v2beta2.go]: answer: we haven't marked any existing reservations yet, so this is OK (but confusing)

**Comment by gandro** [comment]
I have renamed the new IPAM pools mode to `multi-pool`. All related temporary options and files also have been renamed accordingly.

We'll add proper CI and how-to-use-docs once this is properly integrated with multi-network. For now, this is the Helm command I use to test that it works after running `make kind-image`:

```console
helm upgrade --install cilium ./install/kubernetes/cilium --namespace kube-system \
    --set debug.enabled=true \
    --set image.override=localhost:5000/cilium/cilium-dev:local \
    --set image.pullPolicy=IfNotPresent \
    --set operator.image.override=localhost:5000/cilium/operator-generic:local \
    --set operator.image.pullPolicy=IfNotPresent \
    --set operator.replicas=1 \
    --set hubble.relay.enabled=true \
    --set ipam.mode=multi-pool \
    --set "ipam.operator.multiPoolMap.default.ipv4CIDRs[0]=10.0.0.0/8" \
    --set "ipam.operator.multiPoolMap.default.ipv4MaskSize=27" \
    --set "extraConfig.multi-pool-node-pre-alloc=default=64" \
    # Limitation: MultiPool does not support tunneling mode, use direct routing instead
    --set tunnel=disabled \
    --set autoDirectNodeRoutes=true \
    --set ipv4NativeRoutingCIDR=10.0.0.0/8 \
    # Limitation: MultiPool does not support local node routes, use endpoint routes instead
    --set endpointRoutes.enabled=true \
    --set-string extraConfig.enable-local-node-route=false \
    # Limitation: MultiPool does not legacy masquerading, use BPF-based masquerading instead (c.f. cilium/cilium#22273)
    --set kubeProxyReplacement=strict \
    --set bpf.masquerade=true
```

File: operator/cmd/flags.go [hunk 5]
```diff
@@ -39,6 +39,11 @@ func init() {
 		"EC2 Instance tags in the form of k1=v1,k2=v2 (multiple k/v pairs can also be passed by repeating the CLI flag")
 	option.BindEnv(Vp, operatorOption.IPAMInstanceTags)
 
+	flags.Var(option.NewNamedMapOptions(operatorOption.IPAMMultiPoolMap, &operatorOption.Config.IPAMMultiPoolMap, nil), operatorOption.IPAMMultiPoolMap,
+		"IP pool definitions in the form <pool>=ipv4-cidrs:<cidr>,[<cidr>...];ipv4-mask-size:<size> (multiple pools van also be passed by repeating the CLI flag)")
```
Comment (tklauser) [inline operator/cmd/flags.go]: Nit:

```suggestion
		"IP pool definitions in the form <pool>=ipv4-cidrs:<cidr>,[<cidr>...];ipv4-mask-size:<size> (multiple pools can also be passed by repeating the CLI flag)")
```

File: pkg/ipam/allocator/multipool/pool_allocator.go [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+	if !ok {
+		return fmt.Errorf("cannot allocate from non-existing pool: %s", sourcePool)
+	}
+
+	log.WithFields(logrus.Fields{
+		"targetNode": targetNode,
+		"sourcePool": sourcePool,
+		"family":     family,
+		"toAllocate": toAllocate,
+	}).Debug("allocating cidr")
```
Comment (tklauser) [inline pkg/ipam/allocator/multipool/pool_allocator.go:317]: This log line might be confusing because it will also log for negative `toAllocate`:

```
level=debug msg="allocating cidr" family=ipv4 sourcePool=default subsys=ipam-allocator-clusterpool-v2 targetNode=kind-worker toAllocate=-26
```

but the loop below will be skipped and nothing allocated. Probably better to skip the log message if `toAllocate <= 0`.

**Review by tklauser — APPROVED** [review]
LGTM, two small remarks inline.

Also verified using the command given in https://github.com/cilium/cilium/pull/22762#issuecomment-1543938640 that the pool get created properly:

```
% ks logs cilium-operator-5dd5f6dd8c-5bkvl
[...]
level=info msg="  --multi-pool-map='default=ipv4-cidrs:10.0.0.0/8;ipv4-mask-size:27'" subsys=cilium-operator-generic
[...]
level=debug msg="found pool definition" pool="ipv4-cidrs:10.0.0.0/8;ipv4-mask-size:27" pool-name=default subsys=ipam-allocator-clusterpool-v2
level=debug msg="adding pool" ipv4-cidrs="[10.0.0.0/8]" ipv4-masksize=27 ipv6-cidrs="[]" ipv6-masksize=0 pool-name=default subsys=ipam-allocator-clusterpool-v2
[...]
level=debug msg="allocating cidr" family=ipv4 sourcePool=default subsys=ipam-allocator-clusterpool-v2 targetNode=kind-worker toAllocate=64
```

File: pkg/ipam/types/types.go [hunk 7]
```diff
@@ -51,6 +51,51 @@ type AllocationIP struct {
...
+type IPAMPoolAllocation struct {
+	// Pool is the name of the IPAM pool backing this allocation
+	//
+	// +kubebuilder:validation:MinLength=1
+	Pool string `json:"pool"`
+
+	// CIDRs contains a list of pod CIDRs currently allocated from this pool
+	//
+	// +optional
+	CIDRs []string `json:"cidrs,omitempty"`
```
Comment (christarazi) [inline pkg/ipam/types/types.go]: We can validate these CIDRs using

```
+kubebuilder:validation:Format=cidr
```

File: pkg/ipam/types/types.go [hunk 8]
```diff
@@ -159,6 +209,22 @@ type IPAMStatus struct {
...
 }
 
+// IPAMPoolRequest is a request from the agent to the operator, indicating how
+// may IPs it requires from a given pool
+type IPAMPoolDemand struct {
+	// IPv4Addrs contains the number of requested IPv4 addresses out of a given
+	// pool
+	//
+	// +optional
+	IPv4Addrs int `json:"ipv4-addrs,omitempty"`
```
Comment (christarazi) [inline pkg/ipam/types/types.go:224]: Nit: in CRD YAMLs, we have a preference in the codebase for camelCase

File: pkg/ipam/types/types.go [hunk 9]
```diff
@@ -51,6 +51,51 @@ type AllocationIP struct {
...
+
+type IPAMPoolRequest struct {
+	// Pool is the name of the IPAM pool backing this request
+	//
+	// +kubebuilder:validation:MinLength=1
+	Pool string `json:"pool"`
+
+	// Needed indicates how many IPs out of the above Pool this node requests
+	// from the operator. The operator runs a reconciliation loop to ensure each
+	// node always has enough PodCIDRs allocated in each pool to fulfil the
```
Comment (christarazi) [inline pkg/ipam/types/types.go]: Typo: `fulfill`

File: go.mod [hunk 10]
```diff
@@ -92,6 +92,7 @@ require (
 	go.uber.org/goleak v1.2.1
 	go.uber.org/multierr v1.11.0
 	go.universe.tf/metallb v0.11.0
+	go4.org/netipx v0.0.0-20230303233057-f1b76eb4bb35
```
Comment (christarazi) [inline go.mod:94]: TIL: This seems like a nice package that we can use throughout and remove the duplicated functions in `pkg/ip`. Makes more sense as a followup.

File: pkg/ipam/allocator/multipool/pool.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+		if alloc.IsFull() {
+			continue
+		}
+
+		ipnet, err := alloc.AllocateNext()
+		if err != nil {
+			return netip.Prefix{}, err
+		}
+
+		prefix, ok := netipx.FromStdIPNet(ipnet)
```
Comment (christarazi) [inline pkg/ipam/allocator/multipool/pool.go:33]: JFYI and related to my previous comment about this package: this is the only remaining function AFAICT that's missing from our helpers in `pkg/ip`.

File: pkg/ipam/allocator/multipool/pool.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+		if !alloc.InRange(ipnet) {
+			continue
+		}
+
+		allocated, err := alloc.IsAllocated(ipnet)
+		if err != nil {
+			return err
+		}
+		if !allocated {
+			return nil // not an error to release a cidr twice
```
Comment (christarazi) [inline pkg/ipam/allocator/multipool/pool.go:78]: Can this actually occur or is this more of defensive coding?

If it can occur, I'm curious to understand what might trigger more than one release.

File: pkg/ipam/multipool.go [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+	c.waitForAllPools()
+
+	return c
+}
+
+func (m *multiPoolManager) waitForAllPools() {
+	ctx := context.Background()
+	for pool := range m.preallocMap {
+		if m.conf.IPv4Enabled() {
+			m.waitForPool(ctx, IPv4, pool)
```
Comment (christarazi) [inline pkg/ipam/multipool.go]: This context isn't effective. Is this intended?

**Review by christarazi — COMMENTED** [review]
Mostly LGTM, nice work! I left a few minor comments.

I have some questions as well:
1. Should we just announce (and document) deprecation of clusterpool-v2beta?
2. Should we document the known limitations of multi-pool as described in the PR description?
3. I don't see any end-to-end tests and I presume that you & Tobias have manually validated the feature locally. I'd like to see coverage of it sooner rather than later so that we can iron out any bugs before it's released. I wonder if we should configure the kind workflow to run with this new IPAM mode. What are your thoughts on the e2e testing?

File: pkg/ipam/types/types.go (hunk 8 above)
Comment (squeed) [reply pkg/ipam/types/types.go:224]: yeah, camelCase is the convention for k8s yaml / json documents.

**Review by squeed — APPROVED** [review]
Looks great!

File: go.mod (hunk 10 above)
Comment (tklauser) [reply go.mod:94]: Yep, the intention is to replace most of the `netip` convenience funcs in `pkg/ip` by the ones in this package in a follow-up PR (cf. #24246).

**Review by rolinh — APPROVED** [review]
Vendor changes lgtm.

File: pkg/ipam/types/types.go (hunk 8 above)
Comment (gandro) [reply pkg/ipam/types/types.go:224]: Almost all other CiliumNode fields are also `kebab-case` unfortunately, see for example the ENISpec used to manage the ENI IPAM mode: https://github.com/cilium/cilium/blob/c5cbf403dbe355ecbb80dfc8d7a8ed4da45c43bd/pkg/aws/eni/types/types.go#L18

I made `ipv4-addrs` and `ipv6-addrs` kebab-case to be consistent with the existing IPAM fields.

If you feel strongly that we should make newer fields camelCase, I can change it, but otherwise I'll leave it kebab-case.

File: pkg/ipam/allocator/multipool/pool.go (hunk 11 above)
Comment (gandro) [reply pkg/ipam/allocator/multipool/pool.go:33]: Good catch! I'll leave that for a follow-up PR and left a comment here: https://github.com/cilium/cilium/issues/24246#issuecomment-1547961878

File: pkg/ipam/allocator/multipool/pool.go (hunk 12 above)
Comment (gandro) [reply pkg/ipam/allocator/multipool/pool.go:78]: This is defensive coding. I cannot come up with a scenario where we would release an IP twice.

File: pkg/ipam/multipool.go (hunk 13 above)
Comment (gandro) [reply pkg/ipam/multipool.go]: Ah, good catch! The current _behavior_ is intended, i.e. IPAM will block the agent forever if it is not ready. This is consistent with other IPAM modes.

But I think it still makes sense to have a per-pool context, such that we get an error message for each pool which is not ready. I'll fix that.

**Comment by gandro** [comment]
@christarazi Thanks for the review! :pray:  I believe I addressed or commented on all of your feedback. Please take a look.


> I have some questions as well:
> 
>     1. Should we just announce (and document) deprecation of clusterpool-v2beta?
>
>     2. Should we document the known limitations of multi-pool as described in the PR description?
>
>     3. I don't see any end-to-end tests and I presume that you & Tobias have manually validated the feature locally. I'd like to see coverage of it sooner rather than later so that we can iron out any bugs before it's released. I wonder if we should configure the kind workflow to run with this new IPAM mode. What are your thoughts on the e2e testing?

The short answer to all of those questions are that we're planning this in a follow-up PR. The longer answer is that this PR is not really usable as is. Yes, one can get it to run using my [Helm instructions](https://github.com/cilium/cilium/pull/22762#issuecomment-1543938640) above, but there is currently actually no way to have a pod allocate from a pool other than `default`. That was originally planned to be added as part of the multi-homing work, but will now instead likely follow in a separate API call (see [Slack discussion](https://cilium.slack.com/archives/C2B917YHE/p1683892314962849)). In addition, the CRDs will also be added.

So in short, this PR intentionally lacks the ability to 

a) configure IPAM pools using CRDs
b) define a policy to determine what IP pool should be used for a particular workload

Both will be added in a follow-up PR. I'd like to avoid spending time on documentation (including the clusterpool-v2beta deprecation) and end-to-end test writing before those two user-visible changes have been added, because they will strongly influence the content and structure of both the documentation and test code. But I do agree, that once we have the above two user interfaces, that we need to write docs and end to end tests for them.

File: pkg/ipam/types/types.go (hunk 7 above)
Comment (gandro) [reply pkg/ipam/types/types.go]: Fixed by introducing a newtype with that validation annotation.

File: pkg/ipam/allocator/multipool/pool_allocator.go (hunk 6 above)
Comment (gandro) [reply pkg/ipam/allocator/multipool/pool_allocator.go:317]: Fixed by adding a `toAllocate <= 0` check in allocate

**Comment by christarazi** [comment]
Sounds good to me, thanks for the clarification!

File: pkg/ipam/types/types.go (hunk 8 above)
Comment (christarazi) [reply pkg/ipam/types/types.go:224]: Oh that's unfortunate. In that case, better to be consistent with kebab-case.

**Review by christarazi — APPROVED** [review]
It would be good to have an issue to track the followups discussed in https://github.com/cilium/cilium/pull/22762#issuecomment-1548144692

**Comment by gandro** [comment]
> It would be good to have an issue to track the followups discussed in [#22762 (comment)](https://github.com/cilium/cilium/pull/22762#issuecomment-1548144692)

Yes, good point. I've created https://github.com/cilium/cilium/issues/25470
